src/censored.jl

```diff
@@ -0,0 +1,368 @@
+
+The density function (or mass function for discrete `d`) of the censored distribution is
+```math
+f(x; d, l, u) = \\begin{cases}
+    P_d(x \\le l), & x = l \\\\
+    f_d(x),         & l < x < u \\\\
+    P_d(x \\ge u), & x = u \\\\
+  \\end{cases},
+```
+where ``f_d(x)`` is the density (or mass) function of `d`, and ``P_d(x \\le b)`` is the
+cumulative distribution function of ``d`` evaluated at ``x = b``.
+If ``X`` is a random variable from ``d``, then `clamp(X, l, u)` is a random variable from
+its censored version. Note that this implies that even if ``d`` is continuous, its censored
+form assigns positive probability to the bounds ``l`` and `u``. Therefore a censored continuous distribution has atoms and is a mixture of a discrete and continuous components.
+
+The lower bound `l` can be finite or `missing` and the upper bound `u` can be finite or
+`missing`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, one or more of the
+following methods should be implemented:
+- `censored(d::D, l::T, u::T) where {T <: Real}`
+- `censored(d::D, ::Missing, u::Real)`
+- `censored(d::D, l::Real, ::Missing)`
+"""
+censored
+function censored(d::UnivariateDistribution, l::T, u::T) where {T<:Real}
+    return Censored(d, l, u)
+end
+function censored(d::UnivariateDistribution, ::Missing, u::Real)
+    return Censored(d, missing, u)
+end
+function censored(d::UnivariateDistribution, l::Real, ::Missing)
+    return Censored(d, l, missing)
+end
+censored(d::UnivariateDistribution, l::Real, u::Real) = censored(d, promote(l, u)...)
+censored(d::UnivariateDistribution, ::Missing, ::Missing) = d
+
+"""
+    Censored
+
+Generic wrapper for a [`censored`](@ref) distribution.
+"""
+struct Censored{
+    D<:UnivariateDistribution,
+    S<:ValueSupport,
+    T<:Real,
+    TL<:Union{T,Missing},
+    TU<:Union{T,Missing},
+} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::TL      # lower bound
+    upper::TU      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T<:Real}
+        _is_non_empty_interval(l, u) ||
+            error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T, T, T}(d, l, u)
+    end
+    function Censored(d::UnivariateDistribution, l::Missing, u::T) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T, Missing, T}(d, l, u)
+    end
+    function Censored(d::UnivariateDistribution, l::T, u::Missing) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T, T, Missing}(d, l, u)
+    end
+end
+
+function censored(d::Censored, l::T, u::T) where {T<:Real}
+    return censored(
+        d.uncensored,
+        ismissing(d.lower) ? l : max(l, d.lower),
```

Comment:
IIRC generally `x === missing` is superior to `ismissing(x)`, analogously to `=== nothing` and `isnothing`.

Reply:
Ah interesting, i was not aware of this. Per the comments on https://discourse.julialang.org/t/ismissing-x-versus-x-missing/52171/15 , it seems this is no longer the case in Julia 1.7 but indeed was in earlier versions, so perhaps we should just use the equivalence checks for now.